### PR TITLE
[MFA-2038] Unsafe code removal.

### PR DESCRIPTION
**Pull request by Artelas**
Related to
https://github.com/auth0/Guardian.swift/pull/105
Removed unsafe code. Adjusted solution to match unit tests by excluding excessive checks.

### CODE REVIEW

File: Guardian/Generators/OneTimePasswordGenerator.swift [hunk 1]
```diff
@@ -82,17 +82,13 @@ struct OneTimePasswordGenerator: TOTP, HOTP {
...
-            hash &= 0x7fffffff
-            hash = hash % UInt32(pow(10, Float(self.parameters.digits)))
-            return hash
-        }
-
+        let length = MemoryLayout<UInt32>.size
+        // digestData.count - 1 will always be >0, because digestData depends on algorythm and can be only 20, 32, or 64.
+        let offset = Int(digestData[digestData.count - 1] & 0x0f)
+        // offset is always <=15, length is always 4, so prefix will always be within digestData count.
+        var hash = digestData.dropFirst(offset).prefix(length).reduce(0, { $0 << 8 | UInt32($1) })
```
Comment (Widcket) [inline Guardian/Generators/OneTimePasswordGenerator.swift:89]: If I'm understanding correctly, what the original code did (from L85 to L89) was roughly the following:

Screenshot 2023-04-04 at 8 06 25 PM

Could you please expand how reducing a UInt32 worth of data (minus the offset) by bit-shifting to the right (times the current item) is equivalent to getting the value by advancing the pointer?

Comment (Widcket) [inline Guardian/Generators/OneTimePasswordGenerator.swift:89]: If I'm understanding correctly, what the original code did (from L85 to L89) was roughly the following:

Screenshot 2023-04-04 at 8 06 25 PM

Could you please expand how reducing the data (after excluding the offset) by bit-shifting to the left is equivalent to pointee of `start.withMemoryRebound(to: UInt32.self, capacity: 1) { $0 }`?

Comment (Artelas) [reply Guardian/Generators/OneTimePasswordGenerator.swift:89]: let value = start.withMemoryRebound(to: UInt32.self, capacity: 1) { $0 }

Here we have start as pointer to array of UInt8. withMemoryRebound interpret this array as single UInt32 number. For example for array of 4 UInt8 [0x12, 0x34, 0x56, 0x78], withMemoryRebound will produce single UInt32: 0x78563412. Which is then converted by constructor UInt32(bigEndian:) to 0x1234567.

.reduce(0, { $0 << 8 | UInt32($1) })

Here on the first iteration we have UInt32 equal to 0. Which is 0x00000000. First we shift it by 8 bytex, which produces the same value and doing OR with first element which is 0x12. resulting in 0x00000012. On the next iteration we shift it by 8 bytes resulting in 0x00001200. After doing OR with the next element 0x34, the result is 0x00001234. By the end of iterations we have 0x12345678, which is equal to result of withMemoryRebound.

Comment (Widcket) [reply Guardian/Generators/OneTimePasswordGenerator.swift:89]: Fantastic, thanks for the clarification!

**Review by Widcket — APPROVED** [review]

**Review by ionutmanolache-okta — APPROVED** [review]

**Comment by Artelas** [comment]
@ionutmanolache-okta. Yes, every time I've tried to run tests on iOS 16, I got runtime fatal error.

**Comment by ObidosDev** [comment]
Greate! When release of new version? :)

**Pull request merged**